File: exercises/08_ml/r/02_classification.R
# 対訳 — Must 02 — 分類（k近傍法: 形態 → Species_short）
# Culmen_Length / Culmen_Depth / Flipper_Length / Body_MassからSpecies_shortを予測し、
# テストデータのaccuracyと混同行列を表示してください（Python版は classification_report）。
library(readr)
library(dplyr)
library(class)

path <- "../../../data/penguins.csv"
df <- read_csv(path, show_col_types = FALSE) |>
  tidyr::drop_na(
    Culmen_Length, Culmen_Depth, Flipper_Length, Body_Mass, Species_short
  ) |>
  mutate(Species_short = as.character(Species_short))

feats <- c("Culmen_Length", "Culmen_Depth", "Flipper_Length", "Body_Mass")
set.seed(123)
n <- nrow(df)
test_idx <- sample.int(n, size = floor(0.3 * n))
train <- df[-test_idx, ]
test <- df[test_idx, ]

# スケールを揃えてから knn（距離ベースのため）
train_x <- scale(train[, feats])
test_x <- scale(
  test[, feats],
  center = attr(train_x, "scaled:center"),
  scale = attr(train_x, "scaled:scale")
)

pred <- knn(
  train = train_x,
  test = test_x,
  cl = train$Species_short,
  k = 5
)

acc <- mean(pred == test$Species_short)
cat("accuracy:", round(acc, 3), "\n")
print(table(truth = test$Species_short, pred = pred))

File: exercises/08_ml/python/02_classification.py
"""【問題】`Culmen_Length` / `Culmen_Depth` / `Flipper_Length` / `Body_Mass`から
`Species_short`を`KNeighborsClassifier()`で予測し、
テストデータのaccuracyとclassification_reportを表示してください。
"""
import pandas as pd
from sklearn.metrics import accuracy_score, classification_report
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier

df = pd.read_parquet("../../../data/penguins.parquet").dropna(
    subset=["Culmen_Length", "Culmen_Depth", "Flipper_Length", "Body_Mass", "Species_short"]  # 欠損を除去
)

# TODO: 特徴量Xと目的変数yを用意
# TODO: train_test_split（test_size=0.3, random_state=123）
# TODO: KNeighborsClassifierで学習・予測し、accuracyとclassification_reportを表示
